docs/developer-guide/extensions/proxy-extensions.md

```diff
@@ -0,0 +1,190 @@
+the `/extensions/<extension-name>` endpoint exposed by Argo CD API
+server. The example above will proxy requests to
+`<apiserver-host>/extensions/httpbin/` to `http://httpbin.org`.
+
+The diagram below illustrates an interaction possible with this
+configuration:
+
+```
+                                              ┌─────────────┐
+                                              │ Argo CD UI  │
+                                              └────┬────────┘
+                                                   │  ▲
+  GET <apiserver-host>/extensions/httpbin/anything │  │ 200 OK
+            + authn/authz headers                  │  │
+                                                   ▼  │
+                                            ┌─────────┴────────┐
+                                            │Argo CD API Server│
+                                            └──────┬───────────┘
+                                                   │  ▲
+                   GET http://httpbin.org/anything │  │ 200 OK
+                                                   │  │
+                                                   ▼  │
+                                             ┌────────┴────────┐
+                                             │Backend Extension│
+                                             └─────────────────┘
+```
+
+Note that Argo CD API Server requires additional HTTP headers to be
+sent in order to enforce if the incoming request is authenticated and
+authorized before being proxied to the Backend Extension. The headers
+are documented below:
+
+#### `Cookie` (*mandatory*)
+
+Argo CD UI keeps the authentication token stored in a cookie
+(`argocd.token`). This value needs to be sent in the `Cookie` header
+so the API server can validate its authenticity.
+
+Example: 
+
+    Cookie: "argocd.token=eyJhbGciOiJIUzI1Ni..."
+
+The entire Argo CD cookie list can also be sent. The API server will
+filter out the `argocd.token` automatically in this case.
+
+
+#### `Argocd-Application-Name` (mandatory)
+
+The logged in user must have read permission in this application name
+in order to be authorized. The header value must follow the format:
+`"<namespace>:<app-name>"`.
+
+Example:
+
+    Argocd-Application-Name: "namespace:app-name"
+
+#### `Argocd-Project-Name` (mandatory)
+
+The logged in user must have access to this project in order to be
+authorized.
+
+Example:
+
+    Argocd-Project-Name: "default"
+
+#### `Argocd-Resource-GVK-Name` (optional)
+
+To provide additional level of validation, the requested resource can
+be sent in this header. In this case the API server will validate that
+this specific resource belongs to the application identified by the
+`Argocd-Application-Name` header.
+
+The value must follow the format:
+
+    "<apiVersion>:<kind>:<metadata.name>"
+
+Example:
+
+    Argocd-Resource-GVK-Name: "apps/v1:Pod:some-pod"
+
+It is also possible to send multiple values in this header separated
+by commas.
+
+## Security
```

Comment:
We probably need a section detailing how, even though the API server validates `Argocd-Application-Name`, `Argocd-Project-Name`, and `Argocd-Resource-GVK-Name`, those validated values are almost meaningless if the backend does not actually use them. As long as the user can find a resource in an application to which they have access, the API server will happily pass the request along to the backend. It's the backend's responsibility to consider those headers and to only return information appropriate for the user's proven level of access.

Reply:
done